state-chain/pallets/cf-flip/src/lib.rs

```diff
@@ -32,29 +32,74 @@ use scale_info::TypeInfo;
 pub use weights::WeightInfo;
 
 use cf_traits::{
-	AccountInfo, Bonding, DeregistrationCheck, FeePayment, FundingInfo, OnAccountFunded, Slashing,
+	AccountInfo, Bonding, CallInfoId, DeregistrationCheck, FeePayment,
+	FeeScalingCallInfoIdentifier, FundingInfo, OnAccountFunded, Slashing,
 };
 pub use imbalances::{Deficit, ImbalanceSource, InternalSource, Surplus};
 pub use on_charge_transaction::FlipTransactionPayment;
 
-use frame_support::{
-	ensure,
-	traits::{Get, Imbalance, OnKilledAccount, SignedImbalance},
-};
-
 use codec::{Decode, Encode, MaxEncodedLen};
 use frame_support::{
+	ensure,
 	pallet_prelude::*,
 	sp_runtime::{
 		traits::{AtLeast32BitUnsigned, MaybeSerializeDeserialize, Saturating, Zero},
-		DispatchError, Permill, RuntimeDebug,
+		DispatchError, FixedPointNumber, FixedU64, Permill, RuntimeDebug,
 	},
+	traits::{Get, Imbalance, OnKilledAccount, SignedImbalance},
 };
 use frame_system::pallet_prelude::*;
 use sp_std::{fmt::Debug, marker::PhantomData, prelude::*};
 
 pub use pallet::*;
 
+#[derive(
+	CloneNoBound,
+	RuntimeDebugNoBound,
+	PartialEqNoBound,
+	EqNoBound,
+	Encode,
+	Decode,
+	MaxEncodedLen,
+	TypeInfo,
+)]
+pub enum PalletConfigUpdate {
+	SetSlashingRate(Permill),
+	// Set fee scaling rate for any calls that are scaled.
+	SetFeeScalingRate(FeeScalingRateConfig),
+}
+
+#[derive(
+	Encode, Decode, TypeInfo, MaxEncodedLen, Clone, Copy, PartialEq, Eq, RuntimeDebug, Default,
+)]
+pub enum FeeScalingRateConfig {
+	ExponentBuffer {
+		buffer: u16,
+		exp_base: FixedU64,
+	},
+	#[default]
+	NoScaling,
+}
+
+impl FeeScalingRateConfig {
+	pub fn scale_fee<Balance: AtLeast32BitUnsigned + Copy>(
+		&self,
+		pre_scaled_fee: Balance,
+		call_count: u16,
+	) -> Balance {
+		match self {
+			FeeScalingRateConfig::ExponentBuffer { buffer, exp_base } => {
+				let multiplier = (*exp_base)
```

Comment:
Just to note here - if you use y = k^x, when k = 1, the result will always be 1
otherwise y=k^x and y = x^k produces similar exponential curves.

Reply:
Yeah I think it makes sense for the exponent to be constant (ie. 1: linear scaling, 2: quadratic scaling, etc), instead of the reverse.

Reply:
k^x means we don't need to use fixed point types at all.

The multiplier is just `call_count ^ x` where is x is: `0` for constant, `1` for linear, `2` for quadratic, etc. I think this is easier to visualize/reason about.